controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataAuthFocusedTests.java

```diff
@@ -0,0 +1,616 @@
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.net.URI;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.junit.Assert.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * There are authorization-related tests elsewhere ({@link SecureStreamMetaDataTests} and
+ * {@link UserSecureStreamMetaDataTests}) too. Here, we have focused authorization tests which test the logic more
+ * comprehensively; Tests here also run much quicker, since we share REST server, auth handler and its configuration
+ * across all tests in this class.
+ *
+ * Note: Since the tests are intended to run using a shared REST server, it is important to ensure that the tests do not
+ * create resources with the same names, as doing so can make the tests indirectly dependent on each other and flaky.
+ */
+public class StreamMetaDataAuthFocusedTests {
+
+    private final static int HTTP_STATUS_OK = 200;
+    private final static int HTTP_STATUS_CREATED = 201;
+    private final static int HTTP_STATUS_NOCONTENT = 204;
+    private final static int HTTP_STATUS_UNAUTHORIZED = 401;
+    private final static int HTTP_STATUS_FORBIDDEN = 403;
+
+    // Suppressing the checkstyle errors below, as we are using a class initializer (a method with @BeforeClass
+    // annotation) for efficiency, and we cannot make these members final.
+
+    @SuppressWarnings("checkstyle:StaticVariableName")
+    private static ControllerService mockControllerService;
+
+    @SuppressWarnings("checkstyle:StaticVariableName")
+    private static RESTServerConfig serverConfig;
+
+    @SuppressWarnings("checkstyle:StaticVariableName")
+    private static RESTServer restServer;
+
+    @SuppressWarnings("checkstyle:StaticVariableName")
+    private static Client client;
+
+    @SuppressWarnings("checkstyle:StaticVariableName")
+    private static File passwordHandlerInputFile;
+
+    // We want to ensure that the tests in this class are run one after another (in no particular sequence), as we
+    // are using a shared server (for execution efficiency). We use this in setup and teardown method initiazers
+    // for ensuring the desired behavior.
+    Lock sequential = new ReentrantLock();
```

Comment:
I'm wondering if this locking is strictly necessary. Would it be sufficient that we use different resource names (scope and stream) to avoid conflicts?

Reply:
I wanted to avoid the possibility of invisible/indirect dependencies among tests and race conditions leading to flaky and hard-to-debug tests. Also, depending on due care taken in naming resources alone don't appear to be sufficient, IMHO. Moreover, the presence of this lock and its usage makes it visibly apparent that the intention is to run the tests in this class one at a time, and that it'll be true even if it were decided that the rest of the tests are run using parallel execution.